core/src/test/java/org/bitcoinj/core/TransactionTest.java

```diff
@@ -305,6 +305,150 @@ public void witnessTransaction() {
                 tx.getHash().toString());
     }
 
+    @Test
+    public void testWitnessSignatureP2WPKH() {
+        // test vector P2WPKH from:
+        // https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki
+        NetworkParameters netParams = TestNet3Params.get();
+
+        Transaction unsignedTx = new Transaction(netParams, HEX.decode("01000000" +  // version
+                "02" +
+                "fff7f7881a8099afa6940d42d1e7f6362bec38171ea3edf433541db4e4ad969f" + "00000000" + "00" + "eeffffff" +
+                "ef51e1b804cc89d182d279655c3aa89e815b1b309fe287d9b2b55d57b90ec68a" + "01000000" + "00" + "ffffffff" +
+                "02" +
+                "202cb20600000000" + "1976a9148280b37df378db99f66f85c95a783a76ac7a6d5988ac" +
+                "9093510d00000000" + "1976a9143bde42dbee7e4dbe6a21b2d50ce2f0167faa815988ac" +
+                "11000000"));
+        assertEquals(2, unsignedTx.getInputs().size());
+        assertEquals(2, unsignedTx.getOutputs().size());
+
+        ECKey key0 = ECKey.fromPrivate(
+                HEX.decode("bbc27228ddcb9209d7fd6f36b02f7dfa6252af40bb2f1cbc7a557da8027ff866"));
+        assertEquals("2103c9f4836b9a4f77fc0d81f7bcb01b7f1b35916864b9476c241ce9fc198bd25432ac",
+                HEX.encode(ScriptBuilder.createOutputScript(key0).getProgram()));
+        ECKey key1 = ECKey.fromPrivate(
+                HEX.decode("619c335025c7f4012e556c2a58b2506e30b8511b53ade95ea316fd8c3286feb9"));
+        assertEquals("025476c2e83188368da1ff3e292e7acafcdb3566bb0ad253f62fc70f07aeee6357",
+                key1.getPublicKeyAsHex());
+
+        TransactionSignature txSig0 = unsignedTx.calculateSignature(0, key0,
+                ScriptBuilder.createOutputScript(key0).getProgram(),
+                Transaction.SigHash.ALL, false);
+        assertEquals("30450221008b9d1dc26ba6a9cb62127b02742fa9d754cd3bebf337f7a55d114c8e5cdd30be022040529b194ba3f9281a99f2b1c0a19c0489bc22ede944ccf4ecbab4cc618ef3ed01",
+                HEX.encode(txSig0.encodeToBitcoin()));
+
+        Script scriptCode = new ScriptBuilder()
+                .data(ScriptBuilder.createOutputScript(LegacyAddress.fromKey(netParams, key1))
```

Comment:
@Chm While trying to integrate your new methods into LocalTransactionSigner, I stumbled upon this line. Can you help with these two questions?

1. Why is it that a P2PKH script (rather than P2WPKH) is created here, with `ScriptBuilder.createOutputScript(LegacyAddress.fromKey(netParams, key1))`? `key1` is used for receiving to a native segwit address (P2WPKH), so I'd expect `SegwitAddress.fromKey()`. But of course with this change the unit test fails, so your code must be right (on the assumption that the BIP example is correct too).

2. Why is the output script wrapped into a data push? This seems unusual, and it also differs from how it is done for the P2PKH spend in line 328. Again, the unit test proves correctness, but is there any mention in the BIP? I read everything carefully and I couldn't find this wrapped construct anywhere.

My ultimate problem is that though everything works in your two unit tests, I cannot get the signature right (likely the sig hash).